core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java

```diff
@@ -0,0 +1,271 @@
+     * which can leverage fast x86 fetch-and-add instructions.
+     */
+    private static final int INTS_PER_CACHE_LINE = 64 / Integer.BYTES;
+    private static final int USED_OFFSET = INTS_PER_CACHE_LINE;
+    private static final int MIN_USED_OFFSET = INTS_PER_CACHE_LINE * 2;
+    private static final int MAX_USED_OFFSET = INTS_PER_CACHE_LINE * 3;
+    private static final int COUNTERS_INTS = INTS_PER_CACHE_LINE * 4;
+    private final AtomicIntegerArray counters;
+
+
+    public AffinityAwareSessionPool(EventExecutor[] eventExecutors, Supplier<Session> sessionSupplier) {
+        this.sessionSupplier = sessionSupplier;
+        this.agentThreadIdPerExecutor = new IdentityHashMap<>(eventExecutors.length);
+        this.localQueues = new Queue[eventExecutors.length];
+        for (int agentThreadId = 0; agentThreadId < eventExecutors.length; agentThreadId++) {
+            var boxedAgentThreadId = agentThreadId;
+            agentThreadIdPerExecutor.put(eventExecutors[agentThreadId], boxedAgentThreadId);
+        }
+        this.localAgentThreadId = new FastThreadLocal<>() {
+            @Override
+            protected Integer initialValue() {
+                var eventExecutor = ThreadExecutorMap.currentExecutor();
+                if (eventExecutor == null) {
+                    return null;
+                }
+                var agentThreadId = agentThreadIdPerExecutor.get(eventExecutor);
+                if (agentThreadId == null) {
+                    return null;
+                }
+                return agentThreadId;
+            }
+        };
+        this.counters = new AtomicIntegerArray(COUNTERS_INTS);
+    }
+
+    private int nextWorkStealIndex() {
+        return ThreadLocalRandom.current().nextInt(0, localQueues.length);
+    }
+
+    private int nextWorkStealIndex(int excludeIndex) {
+        var localQueuesCount = localQueues.length;
+        var workStealIndex = ThreadLocalRandom.current().nextInt(0, localQueuesCount);
+        if (workStealIndex == excludeIndex) {
+            workStealIndex++;
+            if (workStealIndex == localQueuesCount) {
+                workStealIndex = 0;
+            }
+        }
+        return workStealIndex;
+    }
+
+
+    @Override
+    public Session acquire() {
+        var boxedAgentThreadId = localAgentThreadId.get();
+        if (boxedAgentThreadId == null) {
+            return acquireFromLocalQueues();
+        }
+        var agentThreadId = boxedAgentThreadId.intValue();
+        var localQueues = this.localQueues;
+        var session = localQueues[agentThreadId].poll();
+        if (session != null) {
+            incrementUsed();
+            return session;
+        }
+        if (localQueues.length == 1) {
+            return null;
+        }
+        return acquireFromOtherLocalQueues(localQueues, agentThreadId);
+    }
+
+    private Session acquireFromLocalQueues() {
+        int workStealIndex = nextWorkStealIndex();
+        int currentIndex = workStealIndex;
```

Comment:
```suggestion
        int currentIndex = nextWorkStealIndex();
```

Reply:
thanks, it was a leftover of the previous implamentation, good catch!